File: Week 01/id_181/26.remove-duplicates-from-sorted-array.js

/**
 * 给定一个排序数组，你需要在原地删除重复出现的元素，使得每个元素只出现一次，返回移除后数组的新长度。

不要使用额外的数组空间，你必须在原地修改输入数组并在使用 O(1) 额外空间的条件下完成。

[26. 删除排序数组中的重复项 - 力扣（LeetCode）](https://leetcode-cn.com/problems/remove-duplicates-from-sorted-array/)
*/


/**
 * @param {number[]} nums
 * @return {number}
 */


var removeDuplicates = function(nums) {
  if(nums.length === 0) return 0;
  var i = 0;
  for (var j = 0; j<nums.length;j++) {
      if(nums[i]!==nums[j]) {
          i++;
          if(i!=j) {
              nums[i] = nums[j];
          }
      }
  }
  return i + 1
};

File: Week 01/id_106/26.remove-duplicates-from-sorted-array.java
/*
 * @lc app=leetcode id=26 lang=java
 *
 * [26] Remove Duplicates from Sorted Array
 */

// @lc code=start
class Solution {
    public int removeDuplicates(int[] nums) {
        if (nums.length == 0) return 0; 
        int dup = 0;
        for (int i = 1; i < nums.length; i++){
            if (nums[i] != nums[dup]) {    
                dup++;
                nums[dup] = nums[i];
            }
        }
        return dup+1;
    }
}
// @lc code=end


/*
[0,0,1,1,1,2,2,3,3,4]
[0,1,2,3,4]
1 遍历，记录当前指针，定义一个比当前指针快一步的指针，进行遍历，遇到与当前数不同的数字就复制到当前数字的后面
相同就跨过(前提是数组必须排序)
*/
//2 或者如果允许用另一个数组，当遇到不同的数字就放在另一数组中，直接返回新数组就可以，O(n)
//if (nums.length == 0) return 0; 当数组为空时返回0

